platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java

```diff
@@ -39,58 +47,82 @@
+    private static final String DOWNLOAD_FILE_ACTION = "downloadFile";
+    private static final String GET_FILE_URI_ACTION = "getFileURI";
+    private static final String GET_FILE_URIS_ACTION = "getFileURIs";
+    private static final String DELETE_FILE_ACTION = "deleteFile";
+    private static final String DELETE_FILES_ACTION = "deleteFiles";
 
-        if (wizAssetMan == null) {
-            wizAssetMan = new WizAssetManager(cordova.getActivity().getApplicationContext());
-        }
+    private String pathToStorage;
 
-        if ("downloadFile".equals(action)) {
-            Log.d(TAG, "[downloadFile] *********** " + args.toString() );
+    @Override
+    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
+        super.initialize(cordova, webView);
 
-            PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
-            result.setKeepCallback(true);
-            callbackContext.sendPluginResult(result);
+        final Context applicationContext = cordova.getActivity().getApplicationContext();
+        pathToStorage = applicationContext.getCacheDir().getAbsolutePath() + File.separator;
+        wizAssetManager = new WizAssetManager(applicationContext);
+    }
 
+    @Override
+    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
+        if (action.equals(DOWNLOAD_FILE_ACTION)) {
+            final String url = args.getString(0);
+            final String fileURL = args.getString(1);
             final CallbackContext _callbackContext = callbackContext;
-            final String _url = args.getString(0);
-            final String _storePath = args.getString(1);
-
-            cordova.getActivity().runOnUiThread(new Runnable() {
+            cordova.getThreadPool().execute(new Runnable() {
                 public void run() {
                     // Split by "/"
-                    String[] splitURL = _storePath.split("/");
+                    String[] splitURL = fileURL.split("/");
+                    // Last element is name
+                    String fileName = splitURL[splitURL.length - 1];
 
-                    // Last element is file name
-                    String fileName = splitURL[splitURL.length-1];
-
-                    // Build directory
                     String dirName = "";
-                    for (int i=0; i<splitURL.length-1; i++) {
-                        dirName = dirName+splitURL[i]+"/";
+                    for (int i=0; i < splitURL.length-1; i++) {
+                        dirName = dirName + splitURL[i] + "/";
+                    }
+                    String uri = dirName + fileName;
+                    String filePath = buildAssetFilePathFromUri(uri);
+                    String asset = getAssetFilePathFromUri(uri);
+                    File file = new File(filePath);
+                    if (asset != null && file.exists()) {
```

Comment:
It's faster and less likely to suffer from race conditions if you simply setLastModified and don't do an exists check. If setting last modified fails because the file does not exist, you can move to your else block.